File: apps/cli/src/commands/lint.ts
/**
 * `frick lint [--against <path-to-previous-schema.json>]`
 *
 * - Without `--against`: runs the single-schema linter on the foundation
 *   schema and emits a JSON Lines summary.
 * - With `--against <path>`: loads the previous-schema snapshot from disk
 *   and runs a change-lint against the current foundation schema. Each
 *   finding is emitted as its own JSON Lines record, followed by a summary
 *   record. Exit code is 1 when any finding has severity=breaking.
 */
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import {
  foundationSchema,
  lintSchema,
  lintSchemaChange,
  type FrickLintResult,
  type FrickSchema,
} from "@frick/protocol";
import type { ParsedArgs } from "../argv.js";
import { requireString } from "../argv.js";
import { CliFailureError } from "../errors.js";
import { emit, type OutputOptions } from "../output.js";

export async function lintCommand(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const against = requireString(parsed.flags, "against");
  if (against === undefined) {
    return lintCurrent(out);
  }
  return lintChange(against, out);
}

function lintCurrent(out: OutputOptions): number {
  const result = lintSchema(foundationSchema);
  emitResult(result, out);
  return result.breakingCount > 0 ? 1 : 0;
}

async function lintChange(againstPath: string, out: OutputOptions): Promise<number> {
  let previous: FrickSchema;
  try {
    const raw = await readFile(resolve(againstPath), "utf8");
    previous = JSON.parse(raw) as FrickSchema;
  } catch (error) {
    throw new CliFailureError(
      "lint.previous_unreadable",
      `Could not read previous schema from ${againstPath}: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
  const result = lintSchemaChange(foundationSchema, previous);
  emitResult(result, out);
  return result.breakingCount > 0 ? 1 : 0;
}

function emitResult(result: FrickLintResult, out: OutputOptions): void {
  for (const finding of result.findings) {
    emit(finding, out);
  }
  emit(
    {
      ok: result.breakingCount === 0,
      findings: result.findings.length,
      breaking: result.breakingCount,
    },
    out,
  );
}

File: crates/frick-cli/src/commands/lint.rs
//! `frick lint [--against <prev-schema.json>]`
//! (ported from `apps/cli/src/commands/lint.ts`).
//!
//! Without `--against`: `lint_schema(foundation)`. With `--against <path>`:
//! read the previous schema JSON and `lint_schema_change(current, previous)`.
//! Each finding is emitted as its own JSON line, followed by a summary record
//! `{ok, findings, breaking}`. Exit 1 iff `breaking > 0`.

use std::fs;

use frick_protocol::foundation_schema;
use frick_protocol::schema::FrickSchema;
use frick_schema::{FrickLintResult, lint_schema, lint_schema_change};
use serde_json::json;

use crate::argv::ParsedArgs;
use crate::errors::{CliError, EXIT_FAILURE, EXIT_OK};
use crate::output::Output;

/// `lintCommand`.
pub fn lint_command(parsed: &ParsedArgs, out: &mut Output) -> Result<i32, CliError> {
    match parsed.flag_str("against") {
        None => Ok(lint_current(out)),
        Some(path) => lint_change(path, out),
    }
}

fn lint_current(out: &mut Output) -> i32 {
    let result = lint_schema(&foundation_schema());
    emit_result(&result, out);
    exit_for(&result)
}

fn lint_change(against_path: &str, out: &mut Output) -> Result<i32, CliError> {
    let raw = fs::read_to_string(against_path).map_err(|err| {
        CliError::failure(
            "lint.previous_unreadable",
            format!("Could not read previous schema from {against_path}: {err}"),
        )
    })?;
    let previous: FrickSchema = serde_json::from_str(&raw).map_err(|err| {
        CliError::failure(
            "lint.previous_unreadable",
            format!("Could not read previous schema from {against_path}: {err}"),
        )
    })?;
    let result = lint_schema_change(&foundation_schema(), &previous);
    emit_result(&result, out);
    Ok(exit_for(&result))
}

fn emit_result(result: &FrickLintResult, out: &mut Output) {
    for finding in &result.findings {
        out.emit(&serde_json::to_value(finding).expect("finding is serializable"));
    }
    out.emit(&json!({
        "ok": result.breaking_count == 0,
        "findings": result.findings.len(),
        "breaking": result.breaking_count,
    }));
}

fn exit_for(result: &FrickLintResult) -> i32 {
    if result.breaking_count > 0 {
        EXIT_FAILURE
    } else {
        EXIT_OK
    }
}
